networkx/generators/tests/test_dynamic_networks.py

```diff
@@ -0,0 +1,60 @@
+"""Unit tests for the :mod:`networkx.generators.transport_networks` module."""
+import itertools
+import random
+
+import pytest
+
+import networkx as nx
+
+NODES = range(0, 100)
+EDGES = list(itertools.combinations(NODES, 2))
+STATIC_SCALAR_FIELD_VALUES = {k: random.uniform(5, 25) for k in NODES}
+DYNAMIC_SCALAR_FIELD_VALUES = {
+    k: lambda: v + random.uniform(-5.0, 5.0)
+    for k, v in STATIC_SCALAR_FIELD_VALUES.items()
+}
+
+
+def test_gradient_network_static_values():
```

Comment:
IMO explicit test cases would really help push this forward. Currently, the tests just check that the resulting graph has the same number of nodes as the substrate graph, and that the function gives the same result when called multiple times. It'd be good to have unit tests where there is a clear expected outcome from applying the function to a specific input substrate graph. Perhaps there is a concrete example from one of the references mentioned in the docstring?

Reply:
I agree with you that a concrete test case is needed. I will endeavor to add one.

As to the existing tests, their aim is a bit more complex. The idea is that the substrate graph scalar value attribute can be either a static value or a supplier function. As such, in the static case the result should be the same, always. When the substrate graph is dynamic, the probability that you'll get the same result is for all practical reasons zero on real world graphs. This is what these test check.

Reply:
I don't think there is any way to have the field changing while this gradient network is being calculated. Is that what you are intending?  Do you want the same edge/node to give different distances/values at different points in the algorithm?

Reply:
There is a way, if the value of the scalar field attribute is actually a function. See the 2nd test.

Reply:
Yes -- but the description of `scalar_field_value` says that input could be a function. And I think you mean that the value of the attribute could be a function.  Thank you for pointing to the second test where it is more clear than the doc_string.

Now that I understand a little better what you are thinking of when you say "dynamic", I have many more questions. :)

The scalar values of the nodes in the network change each time you measure them. The function has no input arguments, so while you can have a different function for each node, that node's function must stay the same and must act without any inputs. What dynamics can that simulate besides stationary random values from a fixed distribution? You don't even have a way to know how long it has been since the previous measurement.

This seems quite restrictive to me. And yet it seems quite wild to allow the same node to have a different value every single time it is measured. In what sense then does the node have any value? Don't we need to measure it many times to get some sort of average?  Otherwise the resulting network will not be a gradient network at any time being simulated, and neither would it be some sort of average gradient network.  It's not a snapshot, nor an average over time. It's very cool and somewhat unnerving at the same time. :}